cron.go

```diff
@@ -21,121 +21,121 @@ var (
 
 // CronMask interface exposes a method to check whether the
 // given time.Time matches the expression CronMask was constructed with.
-type CronMask interface {
-	Matches(t time.Time) bool
+type CronMask struct {
+	minute     field
+	hour       field
+	dayOfMonth field
+	month      field
+	dayOfWeek  field
 }
 
-type cronMask struct {
-	Minute     cronField
-	Hour       cronField
-	DayOfMonth cronField
-	Month      cronField
-	DayOfWeek  cronField
+func (c *CronMask) Match(t time.Time) bool {
+	return c.minute.match(t.Minute()) &&
+		c.hour.match(t.Hour()) &&
+		c.dayOfMonth.match(t.Day()) &&
+		c.month.match(int(t.Month())) &&
+		c.dayOfWeek.match(int(t.Weekday()))
 }
 
-type cronField interface {
-	Matches(val int) bool
+type field interface {
+	match(val int) bool
 }
 
-type wildcardCronField struct {
+type wildcard struct {
 }
 
-func (*wildcardCronField) Matches(val int) bool {
+func (wildcard) match(val int) bool {
 	return true
 }
 
-type constantCronField struct {
-	Val int
+type constant struct {
+	val int
 }
 
-func (f *constantCronField) Matches(val int) bool {
-	return f.Val == val
+func (f constant) match(val int) bool {
+	return f.val == val
 }
 
-type rangeCronField struct {
-	Start int
-	End   int
+// range is a reserved keyword
+type rangeF struct {
+	start int
+	end   int
 }
 
-func (f *rangeCronField) Matches(val int) bool {
-	return val >= f.Start && val <= f.End
+func (f rangeF) match(val int) bool {
+	return val >= f.start && val <= f.end
 }
 
-type listCronField struct {
-	Parts []cronField
+type list struct {
+	parts []field
 }
 
-func (f *listCronField) Matches(val int) bool {
-	for _, p := range f.Parts {
-		if p.Matches(val) {
+func (f list) match(val int) bool {
+	for _, p := range f.parts {
+		if p.match(val) {
 			return true
 		}
 	}
 
 	return false
 }
 
-func (c *cronMask) Matches(t time.Time) bool {
-	return c.Minute.Matches(t.Minute()) &&
-		c.Hour.Matches(t.Hour()) &&
-		c.DayOfMonth.Matches(t.Day()) &&
-		c.Month.Matches(int(t.Month())) &&
-		c.DayOfWeek.Matches(int(t.Weekday()))
+func validateRange(i, val int) error {
+	validRange := validRanges[i]
+	if validRange.start <= val && validRange.end >= val {
```

Comment:
this is a new question: reasoning about ranges of time can be tricky. Have you considered the lower closed and upper closed approach? As in [0..1) over integers means 0? I don't know what's better, just contemplating which approach can meet the expectation of more users.

Reply:
mh.. wrong place. The validation is the less relevant place for this pointer, of course.

Reply:
From https://en.wikipedia.org/wiki/Cron#CRON_expression

> Hyphen ( - )
> Hyphens define ranges. For example, 2000–2010 indicates every year between 2000 and 2010, inclusive.

Reply:
I basically tried to drop features that I did not see any use and kept the rest of the spec.

Reply:
I see. My concern comes from that cron was not so much designed primarily for matching moments in time, but rather for defining moments in time. But our primary use case here is defining time ranges. The problem with the both ends inclusive is that it makes it impossible to define precise time ranges, e.g. two time ranges, night and day, without overlapping. I am still not sure if this is a problem, just asking if it was considered.

Reply:
We discussed a bit more regarding this. Having the ranges as `[a, b)` has the advantage of being able to represent time ranges without conflicts meaning `0-30 19 * * *` and `30-59 19 * * *` will both match for a minute at `19:30` compared to `[a,b]`. However, `[a,b]` follows the convention from crontab expressions and a more known format. Diverging from the spec has some risk of making the users learn about these exceptional cases.

In either case, I guess we can wait and see how it feels in general and update accordingly in the future.

Reply:
IMO: the risk is to high. The sub part of the name `cron` with the same syntax, but different behavior is not a good idea.